core/src/main/scala/org/scalafmt/util/TokenOps.scala

```diff
@@ -59,18 +56,32 @@ object TokenOps {
       val newlines = newlinesBetween(tok.between)
       newlines > 1 || (isDocstring(tok.right) && !tok.left.is[Comment])
     } || {
+      val leftOwner = owners(tok.left)
+      val rightOwner = owners(tok.right)
+
       style.newlines.alwaysBeforeTopLevelStatements &&
-        tok.between.count(_.is[KwNew]) < 2 && isTopLevelStatment(tok.right, owners(tok.right))
+        isTopLevelStatment(tok.right, rightOwner) &&
+        !{
+          // keep groups of case class/objects with no body in a block
+          def hasBody(tree: Tree) = tree.parent.map(_.children).getOrElse(Seq.empty).exists {
+            case Template(_, _, _, Some(_)) => true
+            case _ => false
+          }
+
+          (TreeOps.isCaseClass(leftOwner) || TreeOps.isCaseObject(leftOwner)) &&
```

Comment:
Is there a need to distinguish `case class` from `final class`? I think this rule could apply to all `Defn.{Object,Trait,Class}`